stl/src/atomic_wait.cpp

```diff
@@ -331,4 +349,65 @@ __std_atomic_api_level __stdcall __std_atomic_set_api_level(__std_atomic_api_lev
     return _Acquire_wait_functions();
 #endif // !_ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE
 }
+
+#pragma warning(push)
+#pragma warning(disable : 4324) // structure was padded due to alignment specifier
+_Smtx_t* __stdcall __std_atomic_get_mutex(const void* const _Key) noexcept {
+    constexpr size_t _Table_size_power = 8;
+    constexpr size_t _Table_size       = 1 << _Table_size_power;
+    constexpr size_t _Table_index_mask = _Table_size - 1;
+
+    struct alignas(std::hardware_destructive_interference_size) _Table_entry {
+        _Smtx_t _Mutex;
+    };
+
+    static _Table_entry _Table[_Table_size]{};
+
+    auto _Index = reinterpret_cast<const std::uintptr_t>(_Key);
+    _Index ^= _Index >> (_Table_size_power * 2);
+    _Index ^= _Index >> _Table_size_power;
+    return &_Table[_Index & _Table_index_mask]._Mutex;
+}
+#pragma warning(pop)
+
+_NODISCARD unsigned char __stdcall __std_atomic_compare_exchange_128(_Inout_bytecount_(16) long long* _Destination,
+    _In_ long long _ExchangeHigh, _In_ long long _ExchangeLow,
+    _Inout_bytecount_(16) long long* _ComparandResult) noexcept {
+#if !defined(_WIN64)
+    return __std_atomic_compare_exchange_128_fallback(_Destination, _ExchangeHigh, _ExchangeLow, _ComparandResult);
+#elif _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 1
+    return _InterlockedCompareExchange128(_Destination, _ExchangeHigh, _ExchangeLow, _ComparandResult);
+#else // ^^^ _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 1 // _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 0 vvv
+    if (__std_atomic_has_cmpxchg16b()) {
+        return _InterlockedCompareExchange128(_Destination, _ExchangeHigh, _ExchangeLow, _ComparandResult);
+    }
+
+    return __std_atomic_compare_exchange_128_fallback(_Destination, _ExchangeHigh, _ExchangeLow, _ComparandResult);
+#endif // ^^^ _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 0
+}
+
+_NODISCARD bool __stdcall __std_atomic_has_cmpxchg16b() noexcept {
+#if !defined(_WIN64)
+    return false;
+#elif _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 1
+    return true;
+#else // ^^^ _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 1 // _STD_ATOMIC_ALWAYS_USE_CMPXCHG16B == 0 vvv
+    enum class _Cmpxchg16_support : char {
+        _Absent  = false,
+        _Present = true,
+        _Unknown,
+    };

```

Comment:
Should these explicitly be given the values `0`, `1`, and `2`, because converting `false` and `true` to `char` is unusual, and implicitly giving `_Unknown` the value `2` is also somewhat unusual?

Reply:
Performance preference to reinterpret char to bool

Reply:
used `char` to avoid it